### PR TITLE
Provide functionality to notice changes in node tree

### PR DESCRIPTION
**Pull request by KoishiKomeiji**
Minor changes:
Share the PodLocation structure between RCs and DSs
Ask the caller to WatchDesires to send a daemon set when a change has been made so that you can verify it
Fix tests to for the above change to WatchDesires

Major changes:
Add a WatchMatchDiff function for each daemon set process to detect changes in node tree
Provide and modify tests for the above WatchDiff function

### CODE REVIEW

File: pkg/ds/daemon_set.go [hunk 1]
```diff
@@ -161,8 +150,17 @@ func (ds *daemonSet) WatchDesires(
 					continue
 				}
 
-			case <-deletedCh:
-				ds.logger.NoFields().Infof("Received daemon set delete signal")
+			case deleteDS := <-deletedCh:
```
Comment (mpuncel) [inline pkg/ds/daemon_set.go]: is the caller of `WatchDesires()` going to be able to know the contents of the deleted daemon set? I would think all we'll know is the uuid of it

Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: The current implementation `WatchDiff` will allow the caller to know what was deleted because it keeps a copy of the daemon set in memory. Since the farm uses `WatchDiff` and `WatchDesires`, it will know the contents.

Comment (mpuncel) [inline pkg/ds/daemon_set.go]: what happens if a daemon set is deleted while the process isn't running?

Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: Currently, if a daemon set is deleted, we will lose track of it, I plan to include a check in the daemon set farm during initialization that will unschedule nodes that are labelled with a daemon set uuid when their daemon set does not exist in the tree.

File: pkg/ds/daemon_set.go [hunk 2]
```diff
@@ -141,11 +126,15 @@ func (ds *daemonSet) WatchDesires(
 			select {
 			case newDS := <-updatedCh:
 				ds.logger.NoFields().Infof("Received daemon set update signal: %v", newDS)
+				if newDS == nil {
+					ds.logger.NoFields().Infof("Very odd, recieved a nil daemon set during update for %v, recovered", *ds)
```
Comment (anthonybishopric) [inline pkg/ds/daemon_set.go]: Remove "Very odd,"

File: pkg/ds/daemon_set_test.go [hunk 3]
```diff
@@ -101,17 +103,54 @@ func watchDSChanges(
...
 					ds.logger.NoFields().Infof("Watched daemon set get updated: %v", *changedDS)
-					updatedCh <- *changedDS
+					updatedCh <- changedDS
 				}
 			}
 			for _, changedDS := range watched.Deleted {
 				if ds.ID() == changedDS.ID {
-					ds.logger.NoFields().Infof("Watched daemon set get deleted: %v", changedDS)
-					deletedCh <- struct{}{}
+					ds.logger.NoFields().Infof("Watched daemon set get deleted: %v", *changedDS)
```
Comment (anthonybishopric) [inline pkg/ds/daemon_set_test.go]: s/get/was

File: pkg/labels/fake_applicator.go [hunk 4]
```diff
@@ -117,6 +118,79 @@ func (app *fakeApplicator) WatchMatches(selector labels.Selector, labelType Type
 	return ch
 }
 
+func (app *fakeApplicator) WatchDiff(labelType Type, quitCh <-chan struct{}) <-chan *LabeledChanges {
```
Comment (anthonybishopric) [inline pkg/labels/fake_applicator.go]: Why is this necessary? Shouldn't watching a daemon set's node selector be sufficient for detecting node changes?

Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator.go]: Updated, all daemon set processes will now use a WatchMatchDiff function which uses the WatchMatches as a helper function to detect changes.

File: pkg/ds/daemon_set_test.go (hunk 3 above)
Comment (KoishiKomeiji) [inline pkg/ds/daemon_set_test.go]: Updated

File: pkg/ds/daemon_set.go (hunk 2 above)
Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: Updated.

File: pkg/labels/http_applicator.go [hunk 5]
```diff
@@ -95,7 +95,21 @@ func (h *httpApplicator) GetMatches(selector labels.Selector, labelType Type) ([
...
+func (h *httpApplicator) WatchMatchDiff(
+	selector labels.Selector,
+	labelType Type,
+	quitCh <-chan struct{},
+) <-chan *LabeledChanges {
+	panic("Not implemented")
+}
+
+/*
+func (h *httpApplicator) WatchDiff(labelType Type, quitCh <-chan struct{}) <-chan *LabeledChanges {
```
Comment (KoishiKomeiji) [inline pkg/labels/http_applicator.go]: Oops, I'll remove this tomorrow.

File: pkg/labels/consul_applicator.go [hunk 6]
```diff
@@ -259,6 +259,88 @@ func (c *consulApplicator) WatchMatches(selector labels.Selector, labelType Type
...
+			select {
+			case <-quitCh:
+				return
+			case labelsList = <-inCh:
+				if labelsList == nil {
+					c.logger.Errorf("Unexpected nil value received from WatchMatches, recovered")
+					continue
+				}
+			}
+			break
```
Comment (mpuncel) [inline pkg/labels/consul_applicator.go]: rather than using `for {  break }` just do the `select{}` statement

File: pkg/labels/fake_applicator.go [hunk 7]
```diff
@@ -102,21 +104,124 @@ func (app *fakeApplicator) GetMatches(selector labels.Selector, labelType Type)
...
+			select {
+			case <-quitCh:
+				return
+			case labelsList = <-inCh:
+				if labelsList == nil {
+					fmt.Println("Unexpected nil value received from WatchMatches, recovered")
+					continue
+				}
+			}
+			break
```
Comment (mpuncel) [inline pkg/labels/fake_applicator.go]: same

File: pkg/labels/fake_applicator.go [hunk 8]
```diff
@@ -102,21 +104,124 @@ func (app *fakeApplicator) GetMatches(selector labels.Selector, labelType Type)
...
+				if _, ok := oldLabels[id]; !ok {
+					// If it is not observed, then it was created
+					outgoingChanges.Created = append(outgoingChanges.Created, Labeled{
+						ID:        id,
+						LabelType: labelType,
+						Labels:    copySet(labeled.Labels),
+					})
+					oldLabels[id] = labeled
+
+				} else if oldLabels[id].Labels.String() != labeled.Labels.String() {
```
Comment (mpuncel) [inline pkg/labels/fake_applicator.go]: nit: it seems like there's a lot of code-reuse here. Given that it's in a fake implementation I don't care that much but consider extracting this to a function that both real and fake can use

File: pkg/labels/fake_applicator_test.go [hunk 9]
```diff
@@ -2,6 +2,7 @@ package labels
```
Comment (mpuncel) [inline pkg/labels/fake_applicator_test.go]: am I understanding correctly that this is testing `WatchMatchDiff()` on the fake applicator but there's no test for `WatchMatchDiff` on the real one?

Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator_test.go]: Yes.

File: pkg/labels/consul_applicator.go (hunk 6 above)
Comment (KoishiKomeiji) [inline pkg/labels/consul_applicator.go]: Updated

File: pkg/labels/fake_applicator.go (hunk 7 above)
Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator.go]: Updated

File: pkg/labels/fake_applicator.go (hunk 8 above)
Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator.go]: Updated.

File: pkg/labels/fake_applicator_test.go (hunk 9 above)
Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator_test.go]: Updated to add test to the real `WatchMatchDiff`

**Comment by bwester** [comment]
This is a really large PR.  There are a couple problems with large PRs that come to mind when reading through this:
1. When there is more code to review, it takes reviewers longer to understand and consider it all.
2. Uncontroversial changes, like moving `PodLocation` to a different package, get held up by reviewing all the other changes.
3. It's easier to understand one commit that proposes a single logical change instead of several where you implement something one way only to change it later in the same PR. (E.g., 0d2cfe82 and ffbffb16.)

I would like to suggest that you should break this one large PR into several smaller chunks for code review.

**Comment by bwester** [comment]
It's important that your change request include not only the code being changed but the *context* of the change.  Does it solve a problem?  Is there a specific bug it fixes?  Why does this code need changing?  Why is this the right way to change it? What is your intent?

All of this may be obvious to everyone right now, but someone looks back at this code in 6 months, a clear English description will make the code much easier understand.

Please add some context to your git commit messages.

**Comment by KoishiKomeiji** [comment]
Updated, rebased into less commits and added more context to commit messages.

**Comment by bwester** [comment]
From the commit message to 3eff464:
> The tests in WatchDiff have been changed ...
> This will remove the code ...
> This will also change the tests ...
> There are also changes ...
> Also change ...

I want to reiterate that you should work to put one logical change into each commit. Each commit should be a self-contained bundle of changes that accomplish *one* goal and can stand on its own.

Why go to the effort? It's relevant right now because it makes reviewing easier.  When I read through a PR, I have to learn how the code works while also figuring out which parts are related to watching their own nodes, improving type safety, fixing error messages, fixing other bugs, etc.  *You can do that last part for me!* After having written it, you know your own code better than anyone else—you can split apart separate concerns faster and more precisely than I can.

If I don't spend time having to disentangle parts your code, I can spend it checking the code for completeness and correctness. Your review will go faster, you will get higher quality feedback, and everyone will have higher confidence that this code is good. I suspect that failing to do this is one reason why this PR is now a couple weeks old without being approved on an otherwise-fast-moving team.

This isn't simply about the number of commits. Github's UI merges everything together anyway. Instead, it's about you presenting a clear view of your work in a way that makes it understandable to others.

**Comment by bwester** [comment]
:shipit: for commit 7a16654.

You still have the problem that I called out earlier: 3eff464 adds some logging messages that are then changed in b330869. Present it right the first time.

**Comment by KoishiKomeiji** [comment]
Updated, commits are now broken down into separate and more logical parts.

File: pkg/labels/applicator.go [hunk 10]
```diff
@@ -86,5 +92,13 @@ type Applicator interface {
 	// the quit channel - this will be indicated by the closing of the result channel. For
 	// this reason, callers should **always** verify that the channel is closed by checking
 	// the "ok" boolean or using `range`.
-	WatchMatches(selector labels.Selector, labelType Type, quitCh chan struct{}) chan []Labeled
+	WatchMatches(selector labels.Selector, labelType Type, quitCh <-chan struct{}) chan []Labeled
```
Comment (bwester) [inline pkg/labels/applicator.go]: Can you put this change in its own commit or PR?

File: pkg/ds/daemon_set_test.go [hunk 11]
```diff
@@ -33,9 +33,9 @@ func waitForNodes(
 	ds DaemonSet,
 	desired int,
 	desiresErrCh <-chan error,
-	changesErrCh <-chan error,
+	dsChangesErrCh <-chan error,
 ) int {
-	timeout := time.After(1 * time.Second)
+	timeout := time.After(2 * time.Second)
```
Comment (bwester) [inline pkg/ds/daemon_set_test.go]: Why do we need a larger timeout?

Comment (KoishiKomeiji) [inline pkg/ds/daemon_set_test.go]: My test failed due to a timeout issue once, so I'm a little paranoid it might happen in the future.
It doesn't timeout using one second anymore though.

File: pkg/labels/consul_applicator.go [hunk 12]
```diff
@@ -259,6 +259,131 @@ func (c *consulApplicator) WatchMatches(selector labels.Selector, labelType Type
...
+					return
+				}
+			}
+
+			newLabels := make(map[string]Labeled)
+			for _, labeled := range results {
+				newLabels[labeled.ID] = labeled
+			}
+
+			//outgoingChanges := diffLabels(oldLabels, newLabels)
```
Comment (bwester) [inline pkg/labels/consul_applicator.go]: Remove commented code

File: pkg/labels/fake_applicator.go [hunk 13]
```diff
@@ -102,21 +105,73 @@ func (app *fakeApplicator) GetMatches(selector labels.Selector, labelType Type)
...
 	ch := make(chan []Labeled)
+	timer := time.NewTimer(time.Duration(0))
 	go func() {
 		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			// upper bound on request rate
```
Comment (bwester) [inline pkg/labels/fake_applicator.go]: I don't think we need to worry about request rate in `fakeApplicator`

File: pkg/labels/fake_applicator.go [hunk 14]
```diff
@@ -102,21 +105,73 @@ func (app *fakeApplicator) GetMatches(selector labels.Selector, labelType Type)
...
+) <-chan *LabeledChanges {
+	outCh := make(chan *LabeledChanges)
+
+	go func() {
+		defer close(outCh)
+
+		logger := logging.DefaultLogger
+		inCh := app.WatchMatches(selector, labelType, quitCh)
+		diffCh := watchDiffLabels(inCh, quitCh, logger)
+		for {
```
Comment (bwester) [inline pkg/labels/fake_applicator.go]: What does this loop do? As far as I can tell, it copies from one channel to another. Can you just return `diffCh`?

File: pkg/labels/consul_applicator.go [hunk 15]
```diff
@@ -259,6 +259,131 @@ func (c *consulApplicator) WatchMatches(selector labels.Selector, labelType Type
...
+	quitCh <-chan struct{},
+) <-chan *LabeledChanges {
+	outCh := make(chan *LabeledChanges)
+
+	go func() {
+		defer close(outCh)
+
+		inCh := c.WatchMatches(selector, labelType, quitCh)
+		diffCh := watchDiffLabels(inCh, quitCh, c.logger)
+		for {
```
Comment (bwester) [inline pkg/labels/consul_applicator.go]: As in my other comment, what benefit does this loop provide?

File: pkg/labels/fake_applicator.go (hunk 13 above)
Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator.go]: We won't need to worry about it for performance, but I'll need it so the changes are all captured together for the tests to work consistently.

Comment (bwester) [inline pkg/labels/fake_applicator.go]: Unit tests execute in a variety of environments and should not rely on timing for correctness. If you want to make sure that a group of changes are found by this method, you should add different synchronization here.

However, I suspect that what you really want is to test a consumer of the `chan []Labeled`. If so, your unit tests should construct the `[]Labeled` batches you want your consumer to see and send them over a channel you control.

File: pkg/labels/consul_applicator.go (hunk 15 above)
Comment (KoishiKomeiji) [inline pkg/labels/consul_applicator.go]: Updated, it's been removed now.

File: pkg/labels/fake_applicator.go (hunk 14 above)
Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator.go]: Updated.

File: pkg/labels/fake_applicator.go (hunk 13 above)
Comment (KoishiKomeiji) [inline pkg/labels/fake_applicator.go]: Updated, tests are configured to pass without the rate now.

File: pkg/ds/daemon_set.go [hunk 16]
```diff
@@ -97,14 +96,15 @@ func (ds *daemonSet) WatchDesires(
 	quitCh <-chan struct{},
 	updatedCh <-chan *fields.DaemonSet,
 	deletedCh <-chan *fields.DaemonSet,
-	nodesChangedCh <-chan struct{},
 ) <-chan error {
 	errCh := make(chan error)
+	nodesChangedCh := make(chan struct{})
```
Comment (bwester) [inline pkg/ds/daemon_set.go]: Why not `nodesChangedCh := ds.watchNodeChanges(quitCh)` and remove the goroutine you added to the end of the method? Would that accomplish the same thing?

File: pkg/ds/daemon_set.go [hunk 17]
```diff
@@ -205,9 +206,62 @@ func (ds *daemonSet) WatchDesires(
...
+	}()
+
 	return errCh
 }
 
+// Watch for changes to nodes and sends update and delete signals
+func (ds *daemonSet) watchNodeChanges(
+	quitCh <-chan struct{},
+) <-chan struct{} {
+	nodesChangedCh := make(chan struct{})
```
Comment (bwester) [inline pkg/ds/daemon_set.go]: What owns this channel?

File: pkg/ds/daemon_set.go [hunk 18]
```diff
@@ -205,9 +206,62 @@ func (ds *daemonSet) WatchDesires(
...
+				}
+				changes = val
+				if changes == nil {
+					ds.logger.Error("Unexpected nil value received")
+					continue
+				}
+			}
+
+			if len(changes.Created) != 0 || len(changes.Updated) != 0 || len(changes.Deleted) != 0 {
+				nodesChangedCh <- struct{}{}
```
Comment (bwester) [inline pkg/ds/daemon_set.go]: It seems like a wasted opportunity to meticulously gather a list of the created, updated, and deleted nodes in the query, then send a message signaling no more than, "Hey! Something changed!"

Is there any way to make use of the data you have?

File: pkg/ds/daemon_set.go (hunk 17 above)
Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: This function owns the channel, I'll add a
`defer close(nodesChangedCh)`

File: pkg/ds/daemon_set.go (hunk 16 above)
Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: Updated.

File: pkg/ds/daemon_set.go (hunk 17 above)
Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: Updated.

**Comment by bwester** [comment]
I think I'm OK with this PR now.

File: pkg/ds/daemon_set.go (hunk 18 above)
Comment (KoishiKomeiji) [inline pkg/ds/daemon_set.go]: Updated to ds smarter scheduling based on whether there was an create, update, or delete.

File: pkg/ds/daemon_set.go [hunk 19]
```diff
@@ -135,17 +119,26 @@ func (ds *daemonSet) WatchDesires(
...
 			select {
-			case newDS := <-updatedCh:
+			case newDS, ok := <-updatedCh:
+				if !ok {
+					// channel closed
+					return
+				}
 				ds.logger.NoFields().Infof("Received daemon set update signal: %v", newDS)
+				if newDS == nil {
+					ds.logger.NoFields().Fatal("Unexpected a nil daemon set during update", *ds)
```
Comment (petertseng) [inline pkg/ds/daemon_set.go]: can I just ask whether it's OK for this to be fatal? Who will be running it and is it OK with stopping the entire process?

File: pkg/ds/daemon_set.go [hunk 20]
```diff
@@ -161,8 +154,21 @@ func (ds *daemonSet) WatchDesires(
...
-			case <-deletedCh:
-				ds.logger.NoFields().Infof("Received daemon set delete signal")
+			case deleteDS, ok := <-deletedCh:
+				if !ok {
+					// channel closed
+					return
+				}
+				ds.logger.NoFields().Infof("Received daemon set delete signal: %v", deleteDS)
+				if deleteDS == nil {
+					ds.logger.NoFields().Fatal("Unexpected a nil daemon set during delete")
```
Comment (petertseng) [inline pkg/ds/daemon_set.go]: same q about fatal here

Comment (anthonybishopric) [inline pkg/ds/daemon_set.go]: Definitely do not use Fatal - we do not want the entire server to rash.
